### PR TITLE
Add differentiable modulation and demodulation methods for BPSK, QPSK…

### PR DESCRIPTION
**Pull request by selimfirat**
# Making Modulations Differentiable

## Description
This PR introduces differentiable paths in the modulation and demodulation schemes within Kaira. This enables gradient-based training of neural networks that include modulation layers in their architectures.

## Key Features
- Adds new utility functions for differentiable modulation operations in `differentiable.py`
- Extends base modulator and demodulator classes with `forward_soft` methods for differentiable processing
- Implements differentiable modulation for BPSK, QPSK, and QAM schemes
- Adds comprehensive tests for differentiable operations

## Implementation Details
The implementation preserves backward compatibility while adding new capabilities:
- `BaseModulator` and `BaseDemodulator` classes are extended with `forward_soft` methods
- New utility functions in `differentiable.py` provide core operations for differentiable modulation
- The existing `forward` methods remain unchanged
- Each modulation scheme has been updated to properly handle soft bit probabilities

## Testing
Added comprehensive test suite in `tests/modulations/test_differentiable.py` to verify:
- Differentiability of individual operations
- Gradient flow through modulation and demodulation operations
- End-to-end differentiability of the entire modulation/demodulation pipeline

## Use Case Example
```python
# Create modulators and demodulators
modulator = QPSKModulator()
demodulator = QPSKDemodulator()

# Create soft bits with gradients (represents probabilities from a neural network)
soft_bits = torch.tensor([0.1, 0.9, 0.2, 0.8], requires_grad=True)

# Apply differentiable modulation
symbols = modulator.forward_soft(soft_bits)

# Apply channel effects
# ...

# Apply differentiable demodulation
noise_var = 0.1
decoded_bits = demodulator.forward_soft(symbols, noise_var)

# Compute loss and backpropagate gradients
loss = some_loss_function(decoded_bits, target_bits)
loss.backward()  # Gradients will flow through the entire pipeline
```

## Future Work
- Add differentiable implementations for other modulation schemes
- Optimize computational efficiency of differentiable operations
- Provide more examples for neural network training with differentiable modulations

### CODE REVIEW

**Comment by selimfirat** [comment]
Closing PR #10 as it has compatibility issues that cannot be resolved. The feature development will continue in PR #41 with a different approach.
